File: eshop/backend_django/api/serializers/registration.py
import os

from rest_framework import serializers

from api.models import TraderAgreement, TraderBranch, TraderDocument, TraderProfile


class TraderDocumentSerializer(serializers.ModelSerializer):
    trader_code = serializers.CharField(source="trader.trader_code", read_only=True)
    trader_name = serializers.CharField(source="trader.business_name", read_only=True)
    file_url = serializers.SerializerMethodField(read_only=True)
    file_name = serializers.SerializerMethodField(read_only=True)
    file_extension = serializers.SerializerMethodField(read_only=True)

    def get_file_url(self, obj):
        return self._absolute_file_url(obj.file)

    def get_file_name(self, obj):
        return os.path.basename(obj.file.name) if obj.file else ""

    def get_file_extension(self, obj):
        return os.path.splitext(obj.file.name)[1].lower().lstrip(".") if obj.file else ""

    def _absolute_file_url(self, file):
        if not file:
            return None
        url = file.url
        request = self.context.get("request")
        return request.build_absolute_uri(url) if request and url.startswith("/") else url

    class Meta:
        model = TraderDocument
        fields = "__all__"
        read_only_fields = ("uploaded_by", "verified_by", "uploaded_at", "verified_at")


class TraderBranchSerializer(serializers.ModelSerializer):
    trader_code = serializers.CharField(source="trader.trader_code", read_only=True)
    trader_name = serializers.CharField(source="trader.business_name", read_only=True)
    class Meta:
        model = TraderBranch
        fields = "__all__"
        read_only_fields = ("created_at",)


class TraderAgreementSerializer(serializers.ModelSerializer):
    is_active_agreement = serializers.ReadOnlyField()
    trader_code = serializers.CharField(source="trader.trader_code", read_only=True)
    trader_name = serializers.CharField(source="trader.business_name", read_only=True)
    agreement_file_url = serializers.SerializerMethodField(read_only=True)

    def get_agreement_file_url(self, obj):
        if not obj.agreement_file:
            return None
        url = obj.agreement_file.url
        request = self.context.get("request")
        return request.build_absolute_uri(url) if request and url.startswith("/") else url

    class Meta:
        model = TraderAgreement
        fields = "__all__"
        read_only_fields = ("created_by", "created_at", "updated_at")


class TraderProfileListSerializer(serializers.ModelSerializer):
    class Meta:
        model = TraderProfile
        fields = (
            "id", "trader_type", "business_name", "trader_code", "slug", "owner_full_name", "phone", "email",
            "region", "district", "status", "is_verified", "is_featured", "created_at", "updated_at",
        )


class TraderProfileDetailSerializer(serializers.ModelSerializer):
    agreement = TraderAgreementSerializer(read_only=True)
    documents = TraderDocumentSerializer(many=True, read_only=True)
    branches = TraderBranchSerializer(many=True, read_only=True)

    class Meta:
        model = TraderProfile
        fields = "__all__"
        read_only_fields = ("trader_code", "slug", "created_by", "approved_by", "approved_at", "created_at", "updated_at")


class TraderProfileWriteSerializer(serializers.ModelSerializer):
    class Meta:
        model = TraderProfile
        exclude = ("trader_code", "slug", "created_by", "approved_by", "approved_at", "created_at", "updated_at")

File: eshop/admin_web_nextjs/nextjs-admin/src/types/registration.ts
export type TraderType = "company" | "individual";
export type TraderStatus = "pending" | "approved" | "rejected" | "suspended";
export type AgreementStatus = "draft" | "active" | "expired" | "terminated";
export type CommissionType = "percentage" | "fixed" | "none";
export interface TraderProfile { id: number; trader_type: TraderType; business_name: string; slug: string; owner_full_name: string; phone: string; email: string; region: string; district: string; status: TraderStatus; is_verified: boolean; is_featured: boolean; created_at: string; updated_at: string; [key: string]: unknown; }
export interface TraderAgreement { id: number; trader: number; title: string; commission_type: CommissionType; commission_value: string; start_date: string; end_date: string | null; status: AgreementStatus; signed_by_trader: boolean; signed_by_platform: boolean; is_active_agreement?: boolean; [key: string]: unknown; }
export interface TraderDocument { id: number; trader: number; document_type: string; title: string; file: string; verified: boolean; uploaded_at: string; [key: string]: unknown; }
export interface TraderBranch { id: number; trader: number; name: string; phone: string; email: string; region: string; district: string; is_main_branch: boolean; is_active: boolean; [key: string]: unknown; }
